### PR TITLE
Format support

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/3384
Fix https://github.com/Azure/autorest.csharp/issues/3343
Fix https://github.com/Azure/autorest.csharp/issues/3340

- Use typespec diagnostics for typespec compile issues in emitter
- Support string format

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 1]
```diff
@@ -193,6 +208,7 @@ export function isNeverType(type: Type): type is NeverType {
 export function getInputType(
     context: SdkContext,
     type: Type,
+    format: string | undefined,
```
Comment (ArcturusZhang) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:211]: what is the difference between `format?: string` and `format: string | undefined`?
we might need to align with line 109

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 2]
```diff
@@ -463,11 +505,21 @@ export function getInputType(
...
+                const inputType = getInputType(
+                    context,
+                    value.type,
+                    getFormat(program, value),
+                    models,
+                    enums
+                );
+                // if (format && (inputType as InputPrimitiveType)) {
+                //     applyFormat(inputType as InputPrimitiveType, format);
+                // }
```
Comment (ArcturusZhang) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:517]: we should remove this

**Review by ArcturusZhang — COMMENTED** [review]
A refactor suggestion:
I noticed that we are always passing two parameters into some functions: type: Type, format: string | undefined, maybe we could unify these two things and create a new type maybe?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 1 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:211]: @ArcturusZhang  If you define format?: string, it is optional parameter, it must be put in the last one. Here It is not good to pass as the last one.

**Comment by chunyu3** [comment]
> 

Thanks @ArcturusZhang  Do you mean we need to define an type alias, alias stringAndUndefined = string | undefined, and use this for every where we need to apply two types for a parameter?
Yes, if the occurrence of the combine of more types is high, we can do this refactor. for low occurrence, maybe we will not define such alias.

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 2 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts]: removed

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 3]
```diff
@@ -119,7 +124,17 @@ function getCSharpInputTypeKindByIntrinsicModelName(
...
-            return InputTypeKind.String;
+            switch (format?.toLowerCase()) {
+                case "date":
+                    return InputTypeKind.DateTime;
+                case "uri":
+                case "url":
+                    return InputTypeKind.Uri;
+                case "uuid":
+                    return InputTypeKind.Guid;
+                default:
```
Comment (ArcturusZhang) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:135]: can we explicitly handle the case `undefined` and let the `default` branch (not date, uri, url, uuid, or undefined values) to send a warning that this format is unhandled?

**Comment by ArcturusZhang** [comment]
> > 
> 
> Thanks @ArcturusZhang Do you mean we need to define an type alias, alias stringAndUndefined = string | undefined, and use this for every where we need to apply two types for a parameter? Yes, if the occurrence of the combine of more types is high, we can do this refactor. for low occurrence, maybe we will not define such alias.

Well I mean that since we have quite a few methods that has this signature:
```typescript
function doSomething(
 ...,
 type: Type,
 format: string | undefined,
 ...
)
```
and maybe we could have a type or interface like this:
```typescript
interface FormattedType {
  type: Type,
  format: string | undefined
}
```
and those methods could be simplified to
```typescript
function doSomething(
  ...,
  type: FormattedType,
  ...
)
```
and we could have a method to get this thing:
```typescript
function getFormattedType(program: Program, type: Type)
{
return {type: type, format: getFormat(program, type)} as FormattedType;
}
```

I say this because I noticed the pattern when we are calling those method:
```
// on model.ts L387
getInputType(
    context,
    elementType,
    getFormat(program, elementType),
    models,
    enums
),
// on model.ts L401
KeyType: getInputType(
    context,
    key,
    getFormat(program, key),
    models,
    enums
),
ValueType: getInputType(
    context,
    value,
    getFormat(program, value),
    models,
    enums
),
// on model.ts L455
getInputType(
    context,
    dm,
    getFormat(program, dm),
    models,
    enums
);
// and more
```

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 4]
```diff
@@ -119,7 +124,17 @@ function getCSharpInputTypeKindByIntrinsicModelName(
         case "float64":
             return InputTypeKind.Float64;
         case "string":
-            return InputTypeKind.String;
+            switch (format?.toLowerCase()) {
```
Comment (pshao25) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:127]: Does format only applies to string?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 3 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:135]: We need to return `InputTypeKind.String` if format is `undefined` or other invalid format. Invalid format will be ignore. We may not add a new branch, but give out a warning message.

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 4 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:127]: Yes. @format is only applied on string, its string format.

**Comment by chunyu3** [comment]
> ```ts
> FormattedType 
> ```

I like this ideal. I will do this.

File: src/CADL.Extension/Emitter.Csharp/src/type/inputTypeKind.ts [hunk 5]
```diff
@@ -3,31 +3,36 @@
...
     ResourceType = "ResourceType",
     Stream = "Stream",
     String = "String",
     Time = "Time",
     Uri = "Uri",
-    UnKnownKind = "UnknownKind",
-    SByte = "int8",
-    Byte = "uint8",
-    Bytes = "bytes"
+    Enum = "Enum",
```
Comment (ArcturusZhang) [inline src/CADL.Extension/Emitter.Csharp/src/type/inputTypeKind.ts:34]: what does enum mean here?

Comment (ArcturusZhang) [reply src/CADL.Extension/Emitter.Csharp/src/type/inputTypeKind.ts:34]: oh, I see this exists before this PR, but I am curious what does this mean?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 6]
```diff
@@ -366,16 +388,37 @@ export function getInputType(
     function getInputTypeForArray(elementType: Type): InputListType {
         return {
             Name: "Array",
-            ElementType: getInputType(context, elementType, models, enums),
+            ElementType: getInputType(
+                context,
+                {
+                    type: elementType,
+                    format: getFormat(program, elementType)
+                } as FormattedType,
```
[the remark this answers is not in the document]
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts]: resloved

**Review by ArcturusZhang — APPROVED** [review]

**Review by lirenhe — APPROVED** [review]

**Comment by lirenhe** [comment]
The code change in regen PR are expected.  @chunyu3,  please follow up with healthInsight team later to decide the right behaviour of the model change

**Pull request merged**